core/src/main/scala/quasar/destination/gbq/GBQDestination.scala

```diff
@@ -0,0 +1,244 @@
+          ResourceError.notAResource(path))
+      }
+
+      for {
+        tableName <- Stream.eval[F, String](tableNameF)
+        schema <- Stream.fromEither[F](ensureValidColumns(columns).leftMap(new RuntimeException(_)))
+        gbqJobConfig = formGBQJobConfig(schema, config, tableName)
+        authCfgJson = config.authCfg.asJson
+        accessToken <- Stream.eval(GBQAccessToken.token(authCfgJson.toString.getBytes("UTF-8")))
+        _ <- Stream.eval((mkDataset(client, accessToken.getTokenValue, config)))
+        eitherloc <- Stream.eval(mkGbqJob(client, accessToken.getTokenValue, gbqJobConfig))
+        _ <- Stream.eval(
+          Sync[F].delay(
+            log.info(s"(re)creating ${config.authCfg.projectId}.${config.datasetId}.${tableName} with schema ${columns.show}")))
+        _ <- eitherloc match {
+            case Right(locationUri) => {
+              upload(client, bytes, locationUri)
+            }
+            case Left(e) =>
+              Stream.eval(Sync[F].delay(ApplicativeError[F, Throwable].raiseError(
+                new RuntimeException(s"No Location URL from returned from job: $e"))))
+          }
+      } yield ()
+  }
+
+  // NB: config overwrites when re-pushing tables
+  // see schemaUpdateOptions section at
+  // https://cloud.google.com/bigquery/docs/reference/rest/v2/Job
+  private def formGBQJobConfig(
+      schema: NonEmptyList[GBQSchema],
+      config: GBQConfig,
+      tableId: String)
+      : GBQJobConfig =
+    GBQJobConfig(
+      "CSV",
+      "1",
+      "true",
+      schema.toList,
+      Some("DAY"), //TODO: make this configurable
+      WriteDisposition("WRITE_TRUNCATE"), //TODO: make this configurable
+      GBQDestinationTable(config.authCfg.projectId, config.datasetId, tableId), 
+      "21600000", //6hrs load job limit
+      "LOAD")
+
+  def mkErrorString(errs: NonEmptyList[ColumnType.Scalar]): String =
+    errs
+      .map(err => s"Column of type ${err.show} is not supported by Avalanche")
+      .intercalate(", ")
+
+  private def ensureValidColumns(columns: NonEmptyList[Column[ColumnType.Scalar]]): Either[String, NonEmptyList[GBQSchema]] =
+    columns.traverse(mkColumn(_)).toEither leftMap { errs =>
+      s"Some column types are not supported: ${mkErrorString(errs)}"
+    }
+
+  private def mkColumn(c: Column[ColumnType.Scalar]): ValidatedNel[ColumnType.Scalar, GBQSchema] =
+    tblColumnToGbq(c.tpe).map(s => GBQSchema(s, c.name))
+
+  private def tblColumnToGbq(ct: ColumnType.Scalar): ValidatedNel[ColumnType.Scalar, String] =
+    ct match {
+        case ColumnType.String => "STRING".validNel
+        case ColumnType.Boolean => "BOOL".validNel
+        case ColumnType.Number => "NUMERIC".validNel
+        case ColumnType.LocalDate => "DATE".validNel
+        case ColumnType.LocalDateTime => "DATETIME".validNel
+        case ColumnType.LocalTime => "TIME".validNel
+        case ColumnType.OffsetTime => "TIMESTAMP".validNel
```

Comment:
All `OffsetX` are `TIMESTAMPS`, right? is this some kind of gbq limitation

Reply:
I believe these are the column types I have to choose from:
https://cloud.google.com/bigquery/docs/schemas#standard_sql_data_types

So I just picked timestamp for all the offsets.

Reply:
I just found this in the bigquery docs:

https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#time_zones

maybe its best to use String?

Reply:
These docs are so confusing. They say
> A TIMESTAMP object represents an absolute point in time, independent of any time zone or convention such as Daylight Savings Time with microsecond precision.

and

> Canonical format
YYYY-[M]M-[D]D[( |T)[H]H:[M]M:[S]S[.DDDDDD]][time zone]

in the same paragraph. So, `TIMESTAMP` is independent of time zone but it has a time zone. Ok.

Best I can tell, they are extending the SQL `TIMESTAMP` type to include a time zone as metadata, which is ignored when performing any operations that assume the type is of SQL type `TIMESTAMP`. This way they support the SQL `TIMESTAMP` type but allow having a time zone stuck onto it.

> Time zones are used when parsing timestamps or formatting timestamps for display. The timestamp value itself does not store a specific time zone, nor does it change when you apply a time zone offset.

Does that mean that if your datetime-formatted string has time zone `+05:00`, parsing that as a `TIMESTAMP` will create a `TIMESTAMP` value in `UTC` that is at that same instant (i.e. subtracting 5 hours)? Also, what does "apply a time zone offset" even mean? I guess we have to go look at the BigQuery SQL temporal functions library.